kuzzleio/kuzzle
# Upgrade script

# Description

Add a `bin/upgrade` script allowing to upgrade from a major version of Kuzzle to another.

This script lays the groundwork for version upgrades, and its architecture should be able to support multi-version migrations.

I've abundantly tested it... but manually. I'll work on unit tests if I find the time. The documentation is written, though.

Architecture principles:
* an `UpgradeContext` object is created by the main script entrypoint, and passed around to migration sub-programs: it log functions (all logged texts are duplicated on the console and in a file), configuration, and a customized version of [inquirer](https://www.npmjs.com/package/inquirer#methods), the CLI module I'm using 
* connectors are singletons meant to access to resources if needed, but without having to re-require informations from users (for now, there is only a connector to elasticsearch)
* the `bin/.upgrades/version` contains directories, each named `v???`, where `???` is the version number of Kuzzle that can be upgraded by the directory. For now, there is only a `v1` directory
* migration scripts should be runnable any number of times without error

# Specifics to the migration from v1 to v2

This migration is nearly all about upgrading from ES5 to ES7, and this means changing almost everything in the storage.

It also handles in-place migration vs. new stack migration.

This script upgrades:
* elasticsearch:
  * aliases
  * indexes not kuzzle-related that shouldn't be upgraded
  * collections in indexes that are not kuzzle-related
  * the destruction of the obsolete metadata mappings (active and deletedAt)
  * uppercases in collection names
* redis
* configuration files

# Other changes

The list of breaking changes on configuration files wasn't exhaustive. I've also reworked that list to make it clearer.

# How to test

Good question. I've tried to think of every scenario, but as usual, we'll likely come across clients with more imagination than us.

I'll try to duplicate client infrastructures and try the script on them to see how it fares. I invite you to do the same, and to try different, possibly intricated scenarios.
